Dvop-2545 terraform grafana ecs infrastructure

autmwatl opened the pull request:
Plan & apply deploy as expected.

Commit:
DVOP-2545 initial commit

Commit:
DVOP-2545 exclude secrets and terraform

Commit:
DVOP-2545 add standard tfr structure

Commit:
DVOP-2545 add secrets, subnets, vpc, r53, cert

Commit:
DVOP-2545 add secrets, subnets, vpc, r53, cert

Commit:
DVOP-2545 move cert to lb

Commit:
DVOP-2545 amend image address, service network config

Commit:
DVOP-2545 add cert config

Commit:
DVOP-2545 add cert config

Commit:
DVOP-2545 amend version requirement to use pessimistic constraint

Commit:
DVOP-2545 amend version requirement to use pessimistic constraint

Commit:
DVOP-2545 use new name var

Commit:
DVOP-2545 add security groups

Commit:
DVOP-2545 add image version, cert name, repo name

Commit:
DVOP-2545 amend security groups

Commit:
DVOP-2545 amend subnets

Commit:
DVOP-2545 amend subnets

sienkin commented on groups/rds/locals.tf:
```diff
@@ -0,0 +1,35 @@
+  placement_subnet_pattern = local.secrets.placement_subnet_pattern
+  placement_vpc_pattern    = local.secrets.placement_vpc_pattern
+
+  automation_subnet_cidrs = values(zipmap(
+    values(data.aws_subnet.automation).*.availability_zone,
+    values(data.aws_subnet.automation).*.cidr_block
+  ))
+  automation_subnet_ids = values(zipmap(
+    values(data.aws_subnet.automation).*.availability_zone,
+    values(data.aws_subnet.automation).*.id
+  ))
```
These aren't referenced anywhere in the terraform, are they required?

sienkin commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,36 @@
+  placement_subnet_pattern = local.secrets.placement_subnet_pattern
+  placement_vpc_pattern    = local.secrets.placement_vpc_pattern
+
+  automation_subnet_cidrs = values(zipmap(
+    values(data.aws_subnet.automation).*.availability_zone,
+    values(data.aws_subnet.automation).*.cidr_block
+  ))
+  automation_subnet_ids = values(zipmap(
+    values(data.aws_subnet.automation).*.availability_zone,
+    values(data.aws_subnet.automation).*.id
+  ))
```
These aren't referenced anywhere in the terraform, are they required?

sienkin commented on groups/ecs/sg.tf:
```diff
@@ -0,0 +1,37 @@
+  description = "Restricts access for ${var.service}-${var.environment} lb artifactory nodes"
+  vpc_id      = data.aws_vpc.placement.id
+
+  ingress {
+    description     = "lb HTTP ingress from admin and concourse CIDRs"
+    from_port       = 80
+    to_port         = 80
+    protocol        = "tcp"
+    cidr_blocks     = local.placement_subnet_cidrs
+    prefix_list_ids = [data.aws_ec2_managed_prefix_list.administration.id]
+  }
+
+  ingress {
+    description     = "lb HTTPS ingress from admin and concourse CIDRs"
+    from_port       = 443
+    to_port         = 443
+    protocol        = "tcp"
+    cidr_blocks     = local.placement_subnet_cidrs
+    prefix_list_ids = [data.aws_ec2_managed_prefix_list.administration.id]
+  }
```
Concourse doesn't need to access Grafana via HTTP/HTTPS. Also you're referencing `local.placement_subnet_cidrs` which would appear to be the subnets that the ECS cluster is being made avaiulable on - so you're allowing it to connect to itself? That isn't necessary.

sienkin commented on groups/ecs/sg.tf:
```diff
@@ -0,0 +1,37 @@
+resource "aws_security_group" "alb_security_group" {
+  name        = "${var.environment}-${var.service}-lb"
+  description = "Restricts access for ${var.service}-${var.environment} lb artifactory nodes"
+  vpc_id      = data.aws_vpc.placement.id
```
Careful with the copy and paste - this isn't `artifactory`

sienkin commented on groups/ecs/ecs.tf:
```diff
@@ -0,0 +1,47 @@
+}
+
+resource "aws_ecs_task_definition" "grafana_task" {
+  family                = "${var.service}-task"
+  network_mode          = "awsvpc"
+  requires_compatibilities = ["FARGATE"]
+  cpu = "256"
+  memory = "512"
+
+  execution_role_arn = aws_iam_role.ecs_execution_role.arn
+  task_role_arn = aws_iam_role.ecs_task_role.arn
```
Parameters should be aligned

sienkin commented on groups/ecs/ecs.tf:
```diff
@@ -0,0 +1,47 @@
+  depends_on = [
+    aws_iam_role.ecs_execution_role,
+  ]
+  desired_count = 1
```
Ideally this would be a variable defaulted to 1 to allow for overrides if needed.

sienkin commented on groups/ecs/ecs.tf:
```diff
@@ -0,0 +1,47 @@
+  ]
+  desired_count = 1
+
+  network_configuration {
+    subnets = local.placement_subnet_ids
+    assign_public_ip = false
+  }
```
Parameters should be aligned

sienkin commented on groups/ecs/lb.tf:
```diff
@@ -0,0 +1,68 @@
+resource "aws_lb" "grafana_lb" {
+  name               = "${var.service}-lb"
+  internal           = false
+  load_balancer_type = "application"
```
`internal` should be `true` - this doesn't need to be an external ALB

sienkin commented on groups/ecs/lb.tf:
```diff
@@ -0,0 +1,68 @@
+  load_balancer_arn = aws_lb.grafana_lb.arn
+  port              = "443"
+  protocol          = "HTTPS"
+  ssl_policy        = "ELBSecurityPolicy-TLS-1-2-2017-01"
```
We should aim to use `ELBSecurityPolicy-TLS13-1-2-2021-06` for the `ssl_policy`

sienkin commented on groups/ecs/profiles/shared-services-eu-west-2/devops1/vars:
The directory this file is placed in doersn't match the environment it is configured for. I.e. it's in a `devops1` directory while specifying it's for `platform` - one of these should be updated accordingly.

sienkin commented on groups/ecs/ecs.tf:
```diff
@@ -0,0 +1,47 @@
+  [
+    {
+      "name": "grafana-container",
+      "image": "${local.image_owner_id}.dkr.ecr.${var.region}.amazonaws.com/${var.image_repository_name}:${var.grafana_image_version}",
+      "essential": true,
```
This will lead to invalid specification depend on which account and region this code is executed against. 

The ECR repository URI should be specificed wholly, as a variable, and not compounded like this as there's only one ECR repository URI for the container image.

Also you also defined `local.grafana_image` for this and didn't use it.

sienkin commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,36 @@
+  dns_zone_name               = local.secrets.dns_zone_name
+  load_balancer_dns_zone_name = local.secrets.load_balancer_dns_zone_name
+
+  image_owner_id = local.secrets.image_owner_id
+  grafana_image = "${local.image_owner_id}.dkr.ecr.${var.region}.amazonaws.com/${var.image_repository_name}:${var.grafana_image_version}"
+
```
See: https://github.com/companieshouse/ci-grafana-terraform/pull/1/files#r1301116289

sienkin commented on groups/ecs/sg.tf:
```diff
@@ -0,0 +1,37 @@
+  description = "Restricts access for ${var.service}-${var.environment} lb artifactory nodes"
+  vpc_id      = data.aws_vpc.placement.id
+
+  ingress {
+    description     = "lb HTTP ingress from admin and concourse CIDRs"
+    from_port       = 80
+    to_port         = 80
+    protocol        = "tcp"
+    cidr_blocks     = local.placement_subnet_cidrs
+    prefix_list_ids = [data.aws_ec2_managed_prefix_list.administration.id]
+  }
```
Is port 80 access required? A HTTP listener has not been defined.

sienkin commented on groups/rds/profiles/shared-services-eu-west-2/devops1/vars:
The placement of this file within the directory structure doesn't match the environment it's specifying. You have the file in a `devops1` directory but the environment is `platform` - one of these should be updated accordingly.

sienkin commented on groups/ecs/data.tf:
```diff
@@ -0,0 +1,75 @@
+data "vault_generic_secret" "secrets" {
+  path = "team-${var.team}/${var.account_name}/${var.region}/${var.environment}/${var.repository_name}"
+}
```
This doesn't follow our tytpical namespacing on Vault secrets and we don't store secrets based on the source repository name.

autmwatl replied:
Which path should be used, please?

sienkin commented on groups/rds/locals.tf:
```diff
@@ -0,0 +1,35 @@
+locals {
+  account_ids = data.vault_generic_secret.account_ids.data
```
`account_ids` isn't referenced anywehre - is it required?

sienkin commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,36 @@
+locals {
+  account_ids = data.vault_generic_secret.account_ids.data
```
See: https://github.com/companieshouse/ci-grafana-terraform/pull/1/files#r1301127597

sienkin commented on groups/rds/data.tf:
```diff
@@ -0,0 +1,63 @@
+  path = "aws-accounts/account-ids"
+}
+
+data "vault_generic_secret" "secrets" {
+  path = "team-${var.team}/${var.account_name}/${var.region}/${var.environment}/${var.repository_name}"
+}
```
See: https://github.com/companieshouse/ci-grafana-terraform/pull/1/files#r1301125858

autmwatl replied:
Which path should be used, please?

sienkin commented on groups/rds/locals.tf:
```diff
@@ -0,0 +1,35 @@
+  automation_subnet_pattern = local.secrets.automation_subnet_pattern
+  automation_vpc_pattern    = local.secrets.automation_vpc_pattern
+
+  certificate_arn             = lookup(local.secrets, "certificate_arn", null)
+  dns_zone_name               = local.secrets.dns_zone_name
+  load_balancer_dns_zone_name = local.secrets.load_balancer_dns_zone_name
```
These aren't required for RDS

sienkin commented on groups/rds/variables.tf:
```diff
@@ -0,0 +1,116 @@
+  type        = string
+}
+
+variable "default_instance_type" {
+  description = "The default instance type to use for instances"
+  default     = "t3.medium"
+  type        = string
+}
```
This variable isn't used anywhere and can be removed.

sienkin commented on groups/rds/variables.tf:
```diff
@@ -0,0 +1,116 @@
+  description = "The set value for retention period of rds export cloudwatch logs, setting the period in days value"
+  default     = 180
+  type        = number
+}
```
This variable isn't being used anywhere and can be removed.

Commit:
DVOP-2545 requested changes

Commit:
DVOP-2545 requested changes

sienkin commented on groups/ecs/sg.tf:
```diff
@@ -3,6 +3,16 @@ resource "aws_security_group" "alb_security_group" {
+    from_port       = 443
+    to_port         = 443
+    protocol        = "tcp"
+    cidr_blocks     = local.web_access
+    prefix_list_ids = [data.aws_ec2_managed_prefix_list.administration.id]
```
The `cidr_blocks` parameter isn't required here. There's no need for the service to need to connect back in to itself via the ALB.

All we need here is the `prefix_list_ids` parameter.

Commit:
DVOP-2545 requested changes

Jata2k commented on groups/ecs/ecs.tf:
```diff
@@ -0,0 +1,47 @@
+  family                   = "${var.service}-task"
+  network_mode             = "awsvpc"
+  requires_compatibilities = ["FARGATE"]
+  cpu                      = "256"
```
I would pull these values out as variables. We can default those variables to these values but we should have that flexibility

Commit:
DVOP-2545 requested changes

Jata2k commented on groups/ecs/sg.tf:
```diff
@@ -0,0 +1,28 @@
+  vpc_id      = data.aws_vpc.placement.id
+
+  ingress {
+    description     = "lb HTTPS ingress from admin and concourse CIDRs"
```
We'll have to update the description here as I don't think the Concourse CIDRs are included in the administration list

Commit:
DVOP-2545 requested changes

Jata2k commented on groups/ecs/ecs.tf:
```diff
@@ -0,0 +1,47 @@
+      ]
+    }
+  ]
+  DEFINITION
```
Does this work? I'd have thought it would need to be left aligned

Jata2k commented on groups/ecs/iam.tf:
```diff
@@ -0,0 +1,39 @@
+resource "aws_iam_role" "ecs_execution_role" {
+  name = "ecs_execution_role"
+
+  assume_role_policy = <<EOF
```
Have a look at [this](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role.html#example-of-using-data-source-for-assume-role-policy#example-of-using-data-source-for-assume-role-policy) and see if you can define the policy content using the data source instead. The output should be the same but the definition is a bit nicer to work with

Commit:
DVOP-2545 requested changes

Jata2k commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,27 @@
+  load_balancer_dns_zone_name = local.secrets.load_balancer_dns_zone_name
+  image_owner_id              = local.secrets.image_owner_id
+  concourse_access_cidrs      = local.secrets.concourse_access_cidrs
+  web_access                  = concat(local.placement_subnet_cidrs, [local.concourse_access_cidrs])
```
Is `web_access` used anywhere?

Jata2k commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,27 @@
+  dns_zone_name               = local.secrets.dns_zone_name
+  load_balancer_dns_zone_name = local.secrets.load_balancer_dns_zone_name
+  image_owner_id              = local.secrets.image_owner_id
+  concourse_access_cidrs      = local.secrets.concourse_access_cidrs
```
Is `concourse_access_cidrs` used anywhere other than `web_access`?

Jata2k commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,27 @@
+  ssl_certificate_arn         = var.ssl_certificate_name == "" ? aws_acm_certificate_validation.certificate[0].certificate_arn : data.aws_acm_certificate.certificate[0].arn
+
+  dns_zone_name               = local.secrets.dns_zone_name
+  load_balancer_dns_zone_name = local.secrets.load_balancer_dns_zone_name
```
Looks like this might not be used

Jata2k commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,27 @@
+
+  dns_zone_name               = local.secrets.dns_zone_name
+  load_balancer_dns_zone_name = local.secrets.load_balancer_dns_zone_name
+  image_owner_id              = local.secrets.image_owner_id
```
I don't think this is either

Commit:
DVOP-2545 requested changes

Commit:
DVOP-2545 requested changes

Commit:
DVOP-2545 requested changes

Jata2k commented on groups/ecs/locals.tf:
```diff
@@ -0,0 +1,23 @@
+  ))
+  placement_subnet_pattern  = local.secrets.placement_subnet_pattern
+  placement_vpc_pattern     = local.secrets.placement_vpc_pattern
+  automation_subnet_pattern = local.secrets.automation_subnet_pattern
```
Though we may need to cater for this later on none of the automation definitions are used. As such we're better off removing them until needed

Commit:
DVOP-2545 remove automation definitions

Commit:
DVOP-2545 remove automation definitions

autmwatl commented:
Applies as expected.

Commit:
DVOP-2545 move domain name to local variable

Commit:
DVOP-2545 move resource names to local variable, add service lb

Commit:
DVOP-2545 additional perms

Commit:
DVOP-2545 move cats to locals

Commit:
DVOP-2545 move cats to locals

Commit:
DVOP-2545 move cats to locals

Commit:
DVOP-2545 increase desired count to 3

Commit:
DVOP-2545 remove iam from service

Commit:
DVOP-2545 add lb target_type

Commit:
DVOP-2545 temporarily remove listener

Commit:
DVOP-2545 temporarily lb and service

Commit:
DVOP-2545 temporarily remove domain

Commit:
DVOP-2545 restore temporarily removed resources

Commit:
DVOP-2545 fix network config to use ID

Commit:
DVOP-2545 remove static hostPort

Commit:
DVOP-2545 set healthcheck to documented endpoint

Commit:
DVOP-2545 add sg specifically for tasks. add ingress from port 80.

Commit:
DVOP-2545 temporarily remove cluster to allow sg to be replaced

Commit:
DVOP-2545 temporarily remove lb to allow sg to be replaced

Commit:
DVOP-2545 temporarily remove resources to allow sg to be replaced

Commit:
DVOP-2545 restore resources

Commit:
DVOP-2545 set permissions resources to *

Commit:
DVOP-2545 increase health check grace period

Commit:
DVOP-2545 add ecr permissions

Commit:
DVOP-2545 change health endpoint

Commit:
DVOP-2545 add logging

Commit:
DVOP-2545 add logging

Commit:
DVOP-2545 change health check path

Commit:
DVOP-2545 add capacity_providers

Commit:
DVOP-2545 add scaling capabilities

Commit:
DVOP-2545 health check and listener

set healthcheck to healthz on port 3000.

add listener redirect from port 80 to port 443.

Commit:
DVOP-2545 remove lb 3000

Commit:
DVOP-2545 fix listener typo

Commit:
DVOP-2545 reduce healthcheck grace period

Commit:
DVOP-2545 amend healthcheck path

Commit:
DVOP-2545 amend healthcheck config

Commit:
DVOP-2545 amend healthcheck config

Commit:
DVOP-2545 amend healthcheck config

Commit:
DVOP-2545 add rds lookup

find the endpoint of the grafana rds instance

Commit:
DVOP-2545 add vars to configure grafana database

set env and secret vars to remove the need for grafana.ini file.

Commit:
DVOP-2545 add db credentials

Commit:
DVOP-2545 add resource prefix cat

Commit:
DVOP-2545 use resource prefix cat

Commit:
DVOP-2545 use resource prefix cat

Commit:
DVOP-2545 add db_engine

Commit:
DVOP-2545 amend service name to reflect ecs

Commit:
DVOP-2545 remove r53

Commit:
DVOP-2545 remove deletion protection

Commit:
DVOP-2545 temporarily remove db for redeployment

Commit:
DVOP-2545 temporarily remove db for redeployment

Commit:
DVOP-2545 temporarily remove db for redeployment

Commit:
DVOP-2545 temporarily remove db for redeployment

Commit:
DVOP-2545 redeploy rds

Commit:
DVOP-2545 amend db_name

Commit:
DVOP-2545 amend db_name

Commit:
DVOP-2545 remove ref to db module

Commit:
DVOP-2545 add ssm:GetParameters

Commit:
DVOP-2545 remove ssm:GetParameters

Commit:
DVOP-2545 add ssm secrets

Commit:
DVOP-2545 update minor version

Commit:
DVOP-2545 reduce healthy threshold

Commit:
DVOP-2545 reduce timeout

sienkin committed:
Update instance count, update task SG to allow healthchecks from the ALB
